go-controller/pkg/ovn/controller/apbroute/master_controller.go

```diff
@@ -100,32 +104,49 @@ func (c *ExternalGatewayMasterController) GetAdminPolicyBasedExternalRouteIPsFor
 
 // updateStatusAPBExternalRoute updates the CR with the current status of the CR instance, including errors captured while processing the CR during its lifetime
 func (c *ExternalGatewayMasterController) updateStatusAPBExternalRoute(policyName string, gwIPs sets.Set[string],
-	processedError error) error {
+	syncError error) error {
 	if gwIPs == nil {
 		// policy doesn't exist anymore, nothing to do
 		return nil
 	}
 
-	resultErr := retry.RetryOnConflict(util.OvnConflictBackoff, func() error {
-		routePolicy, err := c.apbRoutePolicyClient.K8sV1().AdminPolicyBasedExternalRoutes().Get(context.TODO(), policyName, metav1.GetOptions{})
-		if err != nil {
-			if apierrors.IsNotFound(err) {
-				// policy doesn't exist, no need to update status
-				return nil
-			}
-			return err
+	// get object from the informer cache. No need to get the object from the kube-apiserver, since patch
+	// doesn't check resource versions, and no one else can update status message owned by current zone.
+	routePolicy, err := c.mgr.routeLister.Get(policyName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// policy doesn't exist, no need to update status
+			return nil
 		}
-
-		updateStatus(routePolicy, strings.Join(sets.List(gwIPs), ","), processedError)
-
-		_, err = c.apbRoutePolicyClient.K8sV1().AdminPolicyBasedExternalRoutes().UpdateStatus(context.TODO(), routePolicy, metav1.UpdateOptions{})
-		if !apierrors.IsNotFound(err) {
-			return err
+		return err
+	}
+	newMsg := fmt.Sprintf("%s configured external gateway IPs: %s", c.zoneID, strings.Join(sets.List(gwIPs), ","))
+	if syncError != nil {
+		newMsg = fmt.Sprintf("%s %s: %v", c.zoneID, types.APBRouteErrorMsg, syncError.Error())
+	}
+	needsUpdate := true
+	for _, message := range routePolicy.Status.Messages {
+		if message == newMsg {
+			// found previous status
+			needsUpdate = false
+			break
 		}
+	}
+	if !needsUpdate {
 		return nil
-	})
-	if resultErr != nil {
-		return fmt.Errorf("failed to update AdminPolicyBasedExternalRoutes %s status: %v", policyName, resultErr)
+	}
+
+	patchOptions := metav1.PatchOptions{
+		Force:        utilpointer.Bool(true),
+		FieldManager: c.zoneID,
+	}
+	patch := []byte(fmt.Sprintf(`{"apiVersion":"k8s.ovn.org/v1","kind":"AdminPolicyBasedExternalRoute","status":{"lastTransitionTime":"%s","messages":["%s"]}}`, metav1.Now().ToUnstructured().(string), newMsg))
```

Comment:
so I see you doing server side apply and then identifying each manager by zone id, but I dont get if each node is updating the same field? Like are they all updating the same status field? Maybe some example output will help me understand.

Reply:
It is updating a field that is an array defined as a set with a merge patch strategy. So basically each node adds an entry to the list instead of replacing the whole list. I think the purpose is to have a single entry per node but I don't know what is the key of the set. If the key is the whole message, a single node could add multiple entries to the list if the error changes or if we go from error to success, instead of replacing its previous entry.

Reply:
I thought the plan here was to use conditions where the condition type would be unique for each node. Similar to how ANP is doing it I think.

Reply:
this whole patch strategy is a bit tricky, so here is what you will find in the managedFields
```
    managedFields:
    - apiVersion: k8s.ovn.org/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:status: {}
      manager: cluster-manager
      operation: Apply
      subresource: status
      time: "2023-10-25T11:04:16Z"
    - apiVersion: k8s.ovn.org/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:lastTransitionTime: {}
          f:messages:
            'v:"ovn-control-plane configured external gateway IPs: 1.1.1.1"': {}
      manager: ovn-control-plane
      operation: Apply
      subresource: status
      time: "2023-10-25T11:04:16Z"
    - apiVersion: k8s.ovn.org/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:lastTransitionTime: {}
          f:messages:
            'v:"ovn-worker configured external gateway IPs: 1.1.1.1"': {}
      manager: ovn-worker
      operation: Apply
      subresource: status
      time: "2023-10-25T11:04:16Z"
    - apiVersion: k8s.ovn.org/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:lastTransitionTime: {}
          f:messages:
            'v:"ovn-worker2 configured external gateway IPs: 1.1.1.1"': {}
      manager: ovn-worker2
      operation: Apply
      subresource: status
      time: "2023-10-25T11:04:16Z"
```
every field has its manager, which is set by `patchOptions.FieldManager` for patch call, so every node owns its message in the list.

`Server-Side Apply tries to merge fields based on the actor who manages them instead of overruling based on values. This way multiple actors can update the same object without causing unexpected interference.`  https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy, so what happens is every owner has its elements in the messages list (potentially multiple elements, but in this case only 1). Now every patch from that owner is treated as a new state for all the list elements owned by this manager. This way every time we patch messages with only 1 message, all messages owned by a given manager will be replaced with this 1 message.
There are some nice examples in the tests for merge code, e.g. https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/merge/set_test.go#L502-L541.

There is still a problem of cleaning up messages owned by deleted nodes/zones. We could probably make some cleanup on cluster-manager start to check which zones are present, and delete all non-existing zone messages. I can try to add that part once we agree on the general framework here.

Reply:
Looks great! I wonder what will the client display when the user runs the `kubectl get apbexternalroute` command, since each zone will have its own status recorded.

Reply:
Yeah, I wonder the same thing, how it would it look and perform to have 250 nodes (or more) include their message into each resource. Another approach would be "no news, good news" and just add a message for nodes that have had a failure which of course is not the same thing so it would be a compromise.

Reply:
just `get` will only show `status.status` (cumulative), and all messages are only shown with `-o yaml`